arch/cortex-m4/src/mpu.rs

```diff
@@ -116,160 +128,493 @@ impl MPU {
                 } else {
-                    0
+                    // This case means `start` is 0.
+                    let mut ceil = math::closest_power_of_two(size as u32) as usize;
+                    if ceil < 256 {
+                        ceil = 256
+                    }
+                    ceil / 8
                 }
             };
 
             // Once we have a subregion size, we get a region size by
             // multiplying it by the number of subregions per region.
-            let region_size = subregion_size * 8;
+            let underlying_region_size = subregion_size * 8;
+
             // Finally, we calculate the region base by finding the nearest
             // address below `start` that aligns with the region size.
-            let region_start = start - (start % region_size);
+            let underlying_region_start = start - (start % underlying_region_size);
 
-            if region_size + region_start - start < len {
-                // Sanity check that the amount left over space in the region
-                // after `start` is at least as large as the memory region we
-                // want to reference.
-                return None;
+            // If `size` doesn't align to the subregion size, extend it.
+            if size % subregion_size != 0 {
+                size += subregion_size - (size % subregion_size);
             }
-            if len % subregion_size != 0 {
-                // Sanity check that there is some integer X such that
-                // subregion_size * X == len so none of `len` is left over when
-                // we take the max_subregion.
-                return None;
+
+            let end = start + size;
+            let underlying_region_end = underlying_region_start + underlying_region_size;
+
+            // To use subregions, the region must be at least 256 bytes. Also, we need
+            // the amount of left over space in the region after `start` to be at least as
+            // large as the memory region we want to cover.
+            if subregion_size >= 32 && underlying_region_end >= end {
+                // The index of the first subregion to activate is the number of
+                // regions between `region_start` (MPU) and `start` (memory).
+                let min_subregion = (start - underlying_region_start) / subregion_size;
+
+                // The index of the last subregion to activate is the number of
+                // regions that fit in `len`, plus the `min_subregion`, minus one
+                // (because subregions are zero-indexed).
+                let max_subregion = min_subregion + size / subregion_size - 1;
+
+                // Turn the min/max subregion into a bitfield where all bits are `1`
+                // except for the bits whose index lie within
+                // [min_subregion, max_subregion]
+                //
+                // Note: Rust ranges are minimum inclusive, maximum exclusive, hence
```

Comment:
I don't understand this,

Can you explain why you can't use Rust's exclusive range i.e, ```min_subregion..=max_subregion```?